lib/alloc.h

```diff
@@ -20,6 +20,58 @@
 #include "defines.h"
 
 
+#define CALLOC(n, type)        ((type *) calloc(n, sizeof(type)))
+#define XCALLOC(n, type)       ((type *) xcalloc(n, sizeof(type)))
+#define MALLOCARRAY(n, type)   ((type *) mallocarray(n, sizeof(type)))
+#define XMALLOCARRAY(n, type)  ((type *) xmallocarray(n, sizeof(type)))
+
+#define ALLOCA(type)           ALLOCARRAY(1, type)
+#define MALLOC(type)           MALLOCARRAY(1, type)
+#define XMALLOC(type)          XMALLOCARRAY(1, type)
+#define REALLOC(ptr, type)     REALLOCARRAY(ptr, 1, type)
+#define REALLOCF(ptr, type)    REALLOCARRAYF(ptr, 1, type)
+
+#define REALLOCARRAY(ptr, n, type)                                            \
+({                                                                            \
+	__auto_type  p_ = (ptr);                                              \
+                                                                              \
+	static_assert(__builtin_types_compatible_p(typeof(p_), type *), "");  \
+                                                                              \
+	(type *) reallocarray(p_, n, sizeof(type));                           \
+})
+
+#define REALLOCARRAYF(ptr, n, type)                                           \
+({                                                                            \
+	__auto_type  p_ = (ptr);                                              \
+                                                                              \
+	static_assert(__builtin_types_compatible_p(typeof(p_), type *), "");  \
+                                                                              \
+	(type *) reallocarrayf(p_, n, sizeof(type));                          \
+})
+
+#define XREALLOCARRAY(ptr, n, type)                                           \
+({                                                                            \
+	__auto_type  p_ = (ptr);                                              \
+                                                                              \
+	static_assert(__builtin_types_compatible_p(typeof(p_), type *), "");  \
+                                                                              \
+	(type *) xreallocarray(p_, n, sizeof(type));                          \
+})
+
+#define ALLOCARRAY(n, type)                                                   \
+({                                                                            \
+	type    *p_;                                                          \
+	size_t  n_ = (n);                                                     \
+                                                                              \
+	if (n_ != 0 && n_ > SSIZE_MAX / n_)                                   \
```

Comment:
What exactly is the intent here?

Reply:
Avoid overflow due to multiplication.  Although, now that I think, `alloca(3)` can only allocate very little memory, so this test should be far from true.  I'll just remove it (now that I remember, I think I already considered this and removed it, but for some reason, it seems I didn't).